src/webgpu/api/operation/command_buffer/programmable/programmable_state_test.ts

```diff
@@ -0,0 +1,223 @@
+  // then writes the result to a third bind group.
+  createBindingStatePipeline(
+    passType: PassType,
+    groups: BindGroupIndices,
+    algorthim: String = 'a.value - b.value'
+  ): GPUComputePipeline | GPURenderPipeline {
+    switch (passType) {
+      case PassType.Compute:
+        return this.createBindingStateComputePipeline(groups, algorthim);
+      case PassType.Render:
+      case PassType.RenderBundle:
+        return this.createBindingStateRenderPipeline(groups, algorthim);
+    }
+  }
+
+  createBindingStateComputePipeline(
+    groups: BindGroupIndices,
+    algorthim: String = 'a.value - b.value'
+  ): GPUComputePipeline {
+    const wgsl = `[[block]] struct Data {
+        value : i32;
+      };
+
+      [[group(${groups.a}), binding(0)]] var<storage> a : Data;
+      [[group(${groups.b}), binding(0)]] var<storage> b : Data;
+      [[group(${groups.out}), binding(0)]] var<storage, read_write> out : Data;
+
+      [[stage(compute), workgroup_size(1)]] fn main() {
+        out.value = ${algorthim};
+        return;
+      }
+    `;
+
+    return this.device.createComputePipeline({
+      layout: this.device.createPipelineLayout({
+        bindGroupLayouts: [this.bindGroupLayout, this.bindGroupLayout, this.bindGroupLayout],
+      }),
+      compute: {
+        module: this.device.createShaderModule({
+          code: wgsl,
+        }),
+        entryPoint: 'main',
+      },
+    });
+  }
+
+  createBindingStateRenderPipeline(
+    groups: BindGroupIndices,
+    algorthim: String = 'a.value - b.value'
+  ): GPURenderPipeline {
+    const wgslShaders = {
+      vertex: `
+        [[stage(vertex)]] fn vert_main() -> [[builtin(position)]] vec4<f32> {
+          return vec4<f32>(0.5, 0.5, 0.0, 1.0);
+        }
+      `,
+
+      fragment: `
+        [[block]] struct Data {
+          value : i32;
+        };
+
+        [[group(${groups.a}), binding(0)]] var<storage> a : Data;
+        [[group(${groups.b}), binding(0)]] var<storage> b : Data;
+        [[group(${groups.out}), binding(0)]] var<storage, read_write> out : Data;
+
+        [[stage(fragment)]] fn frag_main() -> [[location(0)]] vec4<f32> {
+          out.value = ${algorthim};
+          return vec4<f32>(1.0, 0.0, 0.0, 1.0);
+        }
+      `,
+    };
+
+    return this.device.createRenderPipeline({
+      layout: this.device.createPipelineLayout({
+        bindGroupLayouts: [this.bindGroupLayout, this.bindGroupLayout, this.bindGroupLayout],
+      }),
+      vertex: {
+        module: this.device.createShaderModule({
+          code: wgslShaders.vertex,
+        }),
+        entryPoint: 'vert_main',
+      },
+      fragment: {
+        module: this.device.createShaderModule({
+          code: wgslShaders.fragment,
+        }),
+        entryPoint: 'frag_main',
+        targets: [{ format: 'rgba8unorm' }],
+      },
+      primitive: { topology: 'point-list' },
+    });
+  }
+
+  beginSimplePass(passType: PassType): GPUProgrammablePassEncoder {
```

Comment:
can probably use the `createEncoder` helper here. Unfortunately it looks like it will need to be moved from ValidationTest to GPUTest, but that's probably the right thing to do.

Reply:
Moved this and did all the refactoring that entails. Increased the size of this patch considerably, but I agree that it's better to reuse existing helpers.